### ISSUE TITLE
Setup Max File Size on a File Upload field doesn't return error if file exceed limit

### ISSUE THREAD

**Issue by romainpoirier** [verbb/formie#627]
**Steps to reproduce**
1. Setup a File Upload field on a multi-pages forms
2. Set 1 MB as Max File Size
3. Add a file that exceeds this limit
4. Submit to go to step
5. The file is saved without error and the next step is shown

### Form settings

- Multi-page form: Yes
- Submission Method: Page Reload
- Client-side Validation: Yes
- Custom Form Templates: Yes

**Additional info**
- Plugin version: 1.4.23
- Craft version: 3.7.18.2
- Multi-site: Yes

**Comment by engram-design** [comment]
File size enforcement is only ever done client-side. Do you have "Validate Form on Submit" enabled for your form, as you've noted you have client-side validation enabled?

**Comment by romainpoirier** [comment]
Yes, the `Validate Form on Submit` is enabled.
![enabled](https://user-images.githubusercontent.com/4634192/140056594-4dc644ae-2e51-4c41-a499-0d1ccc891445.png)

**Comment by engram-design** [comment]
Strange, then you should be getting this validation error: 
![image](https://user-images.githubusercontent.com/1221575/140056869-c4571e13-41c8-4b5b-9067-ad1be1b5a32d.png)

**Comment by engram-design** [comment]
Also note this is now enforced server-side too in https://github.com/verbb/formie/commit/9823389f90eb84e73eb135b8fb87b483a228ff88 now that I've figured out how to get the uploaded files (because validation kicks in before they become asset elements), which might help a little. But would still be good to know why this validation rule isn't kicking in for you.

I note your custom templates, so long as you're including the attributes used for hooking into client-size validation https://github.com/verbb/formie/blob/craft-3/src/templates/_special/form-template/fields/file-upload.html#L13-L19

**Comment by romainpoirier** [comment]
Unfortunately I don't get this validation error on server side.

How can I already apply `9823389` to check if this helps?

Sorry, I was not totally right when I said that I'm using custom templates. In fact, I'm rendering the form using `craft.formie.renderForm`, so I guess I'm using it in an untouched native way.

**Comment by romainpoirier** [comment]
Here is an example of how a File Upload field is rendering:
`<input type="file" id="fields-formie-form-75426182887f98e11-square-still" class="fui-input" name="fields[squareStill][]" multiple="multiple" required="" accept=".bmp, .gif, .jfif, .jp2, .jpe, .jpeg, .jpg, .jpx, .pam, .pfm, .pgm, .png, .pnm, .ppm, .svg, .tif, .tiff, .webp, .pdf" data-fui-id="cartoon-forum-project-submission-form2021-square-still" data-size-min-limit="0" data-size-max-limit="1" data-file-limit="0" data-fui-message="" aria-required="true">`
I can see the `data-size-max-limit="1"` is set.

**Comment by romainpoirier** [comment]
I have setup a reduced test case with a form that have only two tabs, with a required `File Upload` field on each. In this case, I get a `File must be smaller than 1 MB.` error as expected. But that's not the case on my complex multi-pages forms with lots of more fields. Any idea if there's something that could wrongly interact with the validation?

**Comment by engram-design** [comment]
To get this fix early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.4.23",
  "...": "..."
}
```

Then run `composer update`. And that's quite strange. Are you able to let me know the field layout for your complex form? Mostly just the types of fields you're using, particularly any Group fields. They _shouldn't_ be affecting each other but something clearly is.

Will be good to have the [import/export](https://github.com/verbb/formie/issues/465) functionality implemented to help with exactly this sort of issue.

**Comment by romainpoirier** [comment]
Thank you. Now I get the validation for File Upload fields. This validation seems mostly done on server-side only.

I notes some others issues:

- Sometimes, if we have several file fields, and that we are invalid (because of a too high size), the others will not be saved and will have to be re-uploaded;
- Randomly, if you have several file fields, and you leave a file required field empty, everything that had been saved in the current page of the multi-steps forms disappears: you have to fill everything again;
- If we return to a page where files have been loaded and upload new ones, the previous ones are not replaced on the server and remain stored. This happens even if `Settings` > `File Uploads` is set to `Delete files`;

Here are the types of fields I use in my field layout:
- Hidden Field
- Single-line Text
- Number
- Checkboxes
- Radio Buttons
- Agree
- Repeater
- Email Address
- Phone Number
- Multi-line Text
- Dropdown
- File Upload
- And sometimes conditions

**Comment by engram-design** [comment]
> Sometimes, if we have several file fields, and that we are invalid (because of a too high size), the others will not be saved and will have to be re-uploaded;

That would be correct, for any error encountered for a page-reload form. This is because you're circling back from the server, and would be ideal to have client-size validation enabled (which we're trying to get to the bottom of).

> Randomly, if you have several file fields, and you leave a file required field empty, everything that had been saved in the current page of the multi-steps forms disappears: you have to fill everything again;

I've been testing with multiple file fields, so that seems strange. Will review.

> If we return to a page where files have been loaded and upload new ones, the previous ones are not replaced on the server and remain stored. This happens even if Settings > File Uploads is set to Delete files;

Where are you setting `Settings > File Uploads is set to Delete files` ?

I'll see what I can do to replicate this.

**Comment by romainpoirier** [comment]
> Where are you setting Settings > File Uploads is set to Delete files ?

I'm setting this on `/admin/formie/forms/edit/{id}#tab-settings`.

Thank you!

**Comment by engram-design** [comment]
Ah right! So that's an unrelated setting in this instance, with what happens when you delete a submission.

Just after some testing:

> Randomly, if you have several file fields, and you leave a file required field empty, everything that had been saved in the current page of the multi-steps forms disappears: you have to fill everything again;

I can't seem to replicate this. Have a look at the below example just in case something is off on my end. From set to Page Reload, and client-side validation is turned off. I've omitted one required file upload field, while also supplying an image over the max limit of another required one. Seems to retain existing submission information.

![image](https://user-images.githubusercontent.com/1221575/140427879-a1c2f472-2765-419e-8ba1-36f373e02c6d.png)

> If we return to a page where files have been loaded and upload new ones, the previous ones are not replaced on the server and remain stored. This happens even if Settings > File Uploads is set to Delete files;

Yep, so that's a tricky one - but managed to work something out for that in https://github.com/verbb/formie/commit/4cda5b7f11e9d6f3d8cba63163a24a167f83b1d9 which you can update to with the same update code above (`composer update`).

Getting back to the original issue of client-side warnings not appearing, I've created a test form with a collection of those field types, but it's not flagging it. I'm wondering if you'd be able to send through a screenshot or two of your fields. You're welcome to send directly to `web at verbb.io` if that's easier.

**Comment by romainpoirier** [comment]
I can't replicate it again after the last updates, even in my complex form setup 🙈. I'll update later if I get strange results again. But for now, I think we can close this ticket. Thank you for your investigation time.

**Comment by engram-design** [comment]
Good to know! Let's keep an eye on it...

**Comment by engram-design** [comment]
Fixed in [1.4.24](https://github.com/verbb/formie/releases/tag/1.4.24)
